### PR TITLE
[move-lang] make sources_shadow_deps part of parsing

### PR DESCRIPTION
**Pull request by mengxu-fb**
`sources_shadow_deps` allows a module definition found in source file
shadow the same module definition found in dependency files. It is an
argument in `move_check`, `move_compile`, and
`move_construct_pre_compiled_lib` and in all these files, we can see the
following common code snippet:

```rust
let pprog = if sources_shadow_deps {
    shadow_lib_module_definitions(pprog)
} else {
    pprog
};
```

Perhaps a better place for this flag is `move_parse`, and more
specifically, its worker function, `parse_program`. And logically it
makes sense as well since the shadowing should be a parser concept.

A second change in this commit is that the meaning of
`sources_shadow_deps` is extended. In `move_parse`, there is a check,
`check_targets_deps_dont_intersect` (the same says all). However, it
creates a bit of inconvenience when compiling a module in a dir with
dependencies within the same dir. E.g., compile Diem.move by giving
`diem-framework/modules` as dependencies. A common usage pattern is from
the move prover: `mvp Diem.move --dependency=diem-framework/modules`.

The proposed change is to change `check_targets_deps_dont_intersect`
such that when `sources_shadow_deps` is True, do not return an error,
instead, remove the intersection of sources and deps from the deps.
Correspondingly, the name of this function is changed to
`ensure_targets_deps_dont_intersect` to better capture the change.

## Motivation

Needed by #8179 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

CI

### CODE REVIEW

File: language/move-lang/src/lib.rs [hunk 1]
```diff
@@ -716,39 +703,51 @@ fn parse_program(
...
-        .map(|s| format!("    {}", s))
-        .collect::<Vec<_>>()
-        .join("\n");
-    Err(anyhow!(
-        "The following files were marked as both targets and dependencies:\n{}",
-        all_files
-    ))
+    if sources_shadow_deps {
+        deps.retain(|fname| !intersection.contains(fname));
+        Ok(())
```
Comment (tnowacki) [inline language/move-lang/src/lib.rs:739]: nit, I would keep this file in guarded style, so `return Ok(())` and drop the `else`

File: language/move-lang/src/lib.rs [hunk 2]
```diff
@@ -716,39 +703,51 @@ fn parse_program(
...
-        ))
+        let pprog = parser::ast::Program {
+            source_definitions,
+            lib_definitions,
+        };
+        let pprog = if sources_shadow_deps {
+            shadow_lib_module_definitions(pprog)
+        } else {
+            pprog
+        };
```
Comment (tnowacki) [inline language/move-lang/src/lib.rs:714]: This is... bad, but nothing should be done about it now. 

My previous work was a hack and now this hack is making me feel a lot worse as time goes on. 

@tzakian, you are likely going to have to insert some sort of similar filtering logic for testing. 

(cc @bob-wilson per our parse tree discussion)

- Eventually, there will likely be another compiler pass between parsing and expansion. 
- Parsing will not generate an AST and will instead make a parse tree
  - This will facilitate things like an auto-formatter and is also nice for other syntactic tooling
  - It will also let us actually parse comments and cleanup the doc comment hackery (its cool hackery but definitely brittle)
- This ParseTree -> AST conversion will be a really good spot to do things like lib shadowing and test filtering
- When that happens this logic should be in that pass maybe and configured with CompilationEnv
- @tzakian, for your test filtering, feel free to shove that logic here, or maybe make a new skeleton pass that does the filtering, and will also do the ParseTree -> AST conversion later. 

TL;DR This is fine for this PR

**Review by tnowacki — APPROVED** [review]
Feels like a decent extension, makes it easy to do things like `build --deps stdlib/ --files stdlib/Vector.move` or other similar things
